cty/value_ops_test.go

```diff
@@ -765,6 +765,770 @@ func TestValueEquals(t *testing.T) {
+			UnknownVal(Number),
+			false,
+		},
+		{
+			StringVal(""),
+			NullVal(DynamicPseudoType),
+			false,
+		},
+		{
+			StringVal(""),
+			NullVal(String),
+			false,
+		},
+		{
+			StringVal(""),
+			UnknownVal(String),
+			false,
+		},
+		{
+			NullVal(DynamicPseudoType),
+			NullVal(DynamicPseudoType),
+			true,
+		},
+		{
+			NullVal(String),
+			UnknownVal(Number),
+			false, // because second operand might eventually be null
+		},
+		{
+			UnknownVal(String),
+			NullVal(Number),
+			false, // because first operand might eventually be null
+		},
+		{
+			UnknownVal(String),
+			UnknownVal(Number),
+			false, // because both operands might eventually be null
+		},
+		{
+			StringVal("hello"),
+			UnknownVal(Number),
+			false, // because no number value -- even null -- can be equal to a non-null string
+		},
+		{
+			UnknownVal(String),
+			NumberIntVal(1),
+			false, // because no string value -- even null -- can be equal to a non-null number
+		},
+		{
+			ObjectVal(map[string]Value{
+				"a": StringVal("a"),
+			}),
+			// A null value is always known
+			ObjectVal(map[string]Value{
+				"a": NullVal(DynamicPseudoType),
+			}),
+			false,
+		},
+		{
+			ObjectVal(map[string]Value{
+				"a": NullVal(DynamicPseudoType),
+			}),
+			ObjectVal(map[string]Value{
+				"a": NullVal(DynamicPseudoType),
+			}),
+			true,
+		},
+		{
+			ObjectVal(map[string]Value{
+				"a": StringVal("a"),
+				"b": UnknownVal(Number),
+			}),
+			// While we have a dynamic type, the different object types should
+			// still compare false
+			ObjectVal(map[string]Value{
+				"a": NullVal(DynamicPseudoType),
+				"c": UnknownVal(Number),
+			}),
+			false,
+		},
+		{
+			ObjectVal(map[string]Value{
+				"a": StringVal("a"),
+				"b": UnknownVal(Number),
+			}),
+			// While we have a dynamic type, the different object types should
+			// still compare false
+			ObjectVal(map[string]Value{
+				"a": DynamicVal,
+				"c": UnknownVal(Number),
+			}),
+			false,
+		},
+		{
+			ObjectVal(map[string]Value{
+				"a": NullVal(DynamicPseudoType),
+			}),
+			ObjectVal(map[string]Value{
+				"a": DynamicVal,
+			}),
+			false,
+		},
+		{
+			ObjectVal(map[string]Value{
+				"a": NullVal(List(String)),
+			}),
+			// While the unknown val does contain dynamic types, the overall
+			// container types can't conform.
+			ObjectVal(map[string]Value{
+				"a": UnknownVal(List(List(DynamicPseudoType))),
+			}),
+			false,
+		},
+		{
+			ObjectVal(map[string]Value{
+				"a": NullVal(List(List(String))),
+			}),
+			ObjectVal(map[string]Value{
+				"a": UnknownVal(List(List(DynamicPseudoType))),
+			}),
+			false,
+		},
+		{
+			ObjectVal(map[string]Value{
+				"a": SetVal([]Value{
+					ObjectVal(map[string]Value{
+						"b": UnknownVal(String),
+					}),
+				}),
+			}),
+			ObjectVal(map[string]Value{
+				"a": SetVal([]Value{
+					ObjectVal(map[string]Value{
+						"b": UnknownVal(String),
+					}),
+				}),
+			}),
+			true,
+		},
+		{
+			ObjectVal(map[string]Value{
```

Comment:
This right here is where I got lost: I expected that the set short-circuit (IsWhollyKnown) in this PR would have caused this test to fail. It did not.

Reply:
And now I'm found: the types don't match, so it's properly returning false.